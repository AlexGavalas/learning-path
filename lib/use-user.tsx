import {
    FC,
    ReactNode,
    useEffect,
    useState,
    createContext,
    useContext,
} from 'react';

import { Session, User, AuthChangeEvent } from '@supabase/supabase-js';

import { supabase } from './supabase';

type UserContextType = {
    user: User | null;
    userLoaded: boolean;
};

export const UserContext = createContext<UserContextType | undefined>(
    undefined
);

const handleAuthChange = async (
    event: AuthChangeEvent,
    session: Session | null
) => {
    await fetch('/api/auth', {
        body: JSON.stringify({ event, session }),
        credentials: 'same-origin',
        headers: new Headers({ 'Content-Type': 'application/json' }),
        method: 'POST',
    });
};

export const UserContextProvider: FC<{ children?: ReactNode }> = ({
    children,
}) => {
    const [userLoaded, setUserLoaded] = useState(true);
    const [user, setUser] = useState<User | null>(null);

    useEffect(() => {
        const session = supabase.auth.session();

        setUser(session?.user ?? null);
        setUserLoaded(true);

        const { data: authListener } = supabase.auth.onAuthStateChange(
            (event, session) => {
                handleAuthChange(event, session);
                setUser(session?.user ?? null);
            }
        );

        return () => {
            authListener?.unsubscribe();
        };
    }, []);

    const value = {
        user,
        userLoaded,
    };

    return (
        <UserContext.Provider value={value}>{children}</UserContext.Provider>
    );
};

export const useUser = () => {
    const context = useContext(UserContext);

    if (context === undefined) {
        throw new Error(`useUser must be used within a UserContextProvider.`);
    }

    return context;
};
