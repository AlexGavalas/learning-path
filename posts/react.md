---
title: 'React material'
date: '2021-11-11'
---

This is a list with material you can use to learn React. It will be updated frequently, so check it from time to time for new stuff.

You can start with the [tutorial from the React team](https://reactjs.org/tutorial/tutorial.html).
Some others tutorials are [this](https://www.react.express/) and [this](https://react-tutorial.app/).

[Here](https://learnreact.design/posts/what-is-react) is a more visual introduction to React.

Places to learn more about React:

-   [Best practices](https://alexkondov.com/tao-of-react/)
-   [Awesome React](https://github.com/enaqx/awesome-react)

### React relevant libraries

To learn about state management with Redux go [here](https://redux.js.org/tutorials/essentials/part-1-overview-concepts).

Check out the React Query library for data fetching [here](https://react-query.tanstack.com/).
