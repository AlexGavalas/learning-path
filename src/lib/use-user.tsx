import {
    FC,
    ReactNode,
    useEffect,
    useState,
    useContext,
    createContext,
} from 'react';

import { Session, User, AuthChangeEvent } from '@supabase/supabase-js';

import { supabase } from '~lib/supabase';

type UserContextType = {
    user: User | null;
    userLoaded: boolean;
};

export const UserContext = createContext<UserContextType | null>(null);

const handleAuthChange = async (
    event: AuthChangeEvent,
    session: Session | null,
) => {
    await fetch('/api/auth', {
        body: JSON.stringify({ event, session }),
        credentials: 'same-origin',
        headers: new Headers({ 'Content-Type': 'application/json' }),
        method: 'POST',
    });
};

export const UserContextProvider: FC<{ children?: ReactNode }> = ({
    children,
}) => {
    const [userLoaded, setUserLoaded] = useState(true);
    const [user, setUser] = useState<User | null>(null);

    useEffect(() => {
        supabase.auth.getSession().then(({ data: { session } }) => {
            setUser(session?.user ?? null);
            setUserLoaded(true);
        });

        const {
            data: { subscription: authListener },
        } = supabase.auth.onAuthStateChange((event, session) => {
            handleAuthChange(event, session);
            setUser(session?.user ?? null);
        });

        return () => {
            authListener.unsubscribe();
        };
    }, []);

    const value = {
        user,
        userLoaded,
    };

    return (
        <UserContext.Provider value={value}>{children}</UserContext.Provider>
    );
};

export const useUser = () => {
    const context = useContext(UserContext);

    if (!context) {
        throw new Error(`useUser must be used within a UserContextProvider.`);
    }

    return context;
};
