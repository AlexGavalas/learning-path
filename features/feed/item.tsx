import { useState, FormEventHandler } from 'react';
import { format } from 'date-fns';

import { Dialog } from '@components/dialog';
import { Button } from '@components/button';
import { Textarea } from '@components/textarea';

interface ListItemProps {
    post: Post;
    isMine: boolean;
    handleDelete: (id: string) => void;
    onPostUpdate: (id: string, newPost: string) => Promise<void>;
}

export const ListItem = ({
    post,
    isMine,
    handleDelete,
    onPostUpdate,
}: ListItemProps) => {
    const [openEditDialog, setOpenEditDialog] = useState(false);
    const { post: postData, name, id, created_at } = post;

    const postDate = format(new Date(created_at), 'dd/MM/yyyy HH:mm');

    const closeDialog = () => setOpenEditDialog(false);

    const updatePost: FormEventHandler<HTMLFormElement> = async (e) => {
        e.preventDefault();

        const newPost = new FormData(e.currentTarget)
            .get('newpost')
            ?.toString();

        if (!newPost) return;

        await onPostUpdate(post.id, newPost);

        closeDialog();
    };
    
    return (
        <li className="list-none py-6 border-0 border-solid last:border-b-0 flex flex-col">
            <p className="border-solid border-0 border-l-2 border-l-teal-500 dark:border-l-yellow-500 pl-4 break-words">
                {postData}
            </p>
            <div className="self-end flex gap-4 items-center">
                {isMine && (
                    <div className="flex gap-2 h-5">
                        <Button
                            variant="danger"
                            onClick={() => handleDelete(id)}
                        >
                            Delete
                        </Button>
                        <Button onClick={() => setOpenEditDialog(true)}>
                            Edit
                        </Button>
                    </div>
                )}
                <p className="italic text-sm ">
                    - {name} at {postDate}
                </p>
            </div>
            <Dialog
                open={openEditDialog}
                size="md"
                onClickOutside={closeDialog}
            >
                <p className="m-0 mb-2">Make your changes</p>
                <form onSubmit={updatePost}>
                    <Textarea name="newpost" defaultValue={post.post} />
                    <div className="flex gap-2 justify-end">
                        <Button onClick={closeDialog} variant="danger">
                            Cancel
                        </Button>
                        <Button>Save</Button>
                    </div>
                </form>
            </Dialog>
        </li>
    );
};
