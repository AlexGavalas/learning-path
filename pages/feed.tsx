import type { NextPage, GetServerSideProps } from 'next';
import { useState, FormEventHandler, useRef } from 'react';
import { intervalToDuration, add, format } from 'date-fns';

import Layout from '../components/layout';
import { Input } from '../components/input';
import { supabase } from '../lib/supabase';
import { redis } from '../lib/redis';
import { useUser } from '../lib/use-user';

interface PostContent {
    post: string;
    name: string;
}

interface Post {
    data: PostContent;
    key: string;
}

interface FeedProps {
    posts: Post[];
    isLoggedIn: boolean;
}

export const getServerSideProps: GetServerSideProps<FeedProps> = async ({
    req,
}) => {
    const { user } = await supabase.auth.api.getUserByCookie(req);

    const { data: redisKeys } = await redis.keys('posts:*');

    const redisData = await Promise.all(
        redisKeys.map((key: string) => redis.get(key))
    );

    const posts: Post[] = redisData
        .map(({ data }, i) => ({
            data: JSON.parse(data) as unknown as PostContent,
            key: redisKeys[i],
        }))
        .sort((a, b) => {
            const [, tsA] = a.key.split(':');
            const [, tsB] = b.key.split(':');

            return tsB - tsA;
        });

    return {
        props: {
            posts,
            isLoggedIn: !!user,
        },
    };
};

const MAX_CHARS = 256;
const EXPIRE_AFTER_SECS = 60 * 60 * 24 * 5;

const Feed: NextPage<FeedProps> = ({ posts: initialPosts, isLoggedIn }) => {
    const [posts, setPosts] = useState<Post[]>(initialPosts);
    const [hasUser, setHasUser] = useState(isLoggedIn);
    const formRef = useRef<HTMLFormElement>(null);

    const { user, userLoaded } = useUser();

    const addNewNote: FormEventHandler<HTMLFormElement> = async (e) => {
        e.preventDefault();

        if (!formRef.current) return;

        const user = supabase.auth.user();

        const post = new FormData(formRef.current).get('post');

        if (!post || !user) return;

        const key = `posts:${Date.now()}`;

        const data = {
            post: post.toString().slice(0, MAX_CHARS),
            name: user.user_metadata.name,
        };

        await redis.set(key, JSON.stringify(data));
        await redis.expire(key, EXPIRE_AFTER_SECS);

        formRef.current.reset();

        // Optimistic state update
        setPosts((prev) => [{ data, key }, ...prev]);
    };

    const userExists = Boolean(userLoaded && (user || hasUser));

    return (
        <Layout>
            <section className="leading-8 text-xl">
                <h2 className="my-8 text-black dark:text-white">Feed</h2>
                <p className="text-black dark:text-white">
                    Here you can post anything you find interesting for everyone
                    to see. Any content will live for{' '}
                    {
                        intervalToDuration({
                            start: new Date(),
                            end: add(new Date(), {
                                seconds: EXPIRE_AFTER_SECS,
                            }),
                        }).days
                    }{' '}
                    days . Use this section for stuff like TIL, etc...
                </p>
                {userExists ? (
                    <form
                        className="relative h-8 flex"
                        onSubmit={addNewNote}
                        ref={formRef}
                    >
                        <Input
                            name="post"
                            placeholder={`Post anything (max ${MAX_CHARS} characters)`}
                            autoComplete="off"
                        />
                        <div className="h-full p-1 absolute right-0 flex gap-2">
                            <button className="rounded border-0 text-black dark:text-white bg-gray-200 hover:bg-gray-300 dark:bg-gray-500 dark:hover:bg-gray-600">
                                Add
                            </button>
                            <button
                                type="button"
                                onClick={() => {
                                    setHasUser(false);
                                    supabase.auth.signOut();
                                }}
                            >
                                Sign out
                            </button>
                        </div>
                    </form>
                ) : (
                    <div className="flex items-center gap-4 h-full">
                        <p className="m-0 text-black dark:text-white">
                            Login to be able to post
                        </p>
                        <button
                            className="h-8 rounded border-0 text-black dark:text-white hover:bg-gray-300 bg-gray-200 dark:bg-gray-500 dark:hover:bg-gray-600"
                            onClick={() => {
                                supabase.auth.signIn(
                                    { provider: 'google' },
                                    {
                                        redirectTo: `${window.location.origin}/feed`,
                                    }
                                );
                            }}
                        >
                            Login with Google
                        </button>
                    </div>
                )}
                <ul className="mt-4">
                    {posts.map(({ data: { post, name }, key }) => {
                        const [, timestamp] = key.split(':');

                        const postDate = format(
                            new Date(+timestamp),
                            'dd/MM/yyyy HH:mm'
                        );

                        return (
                            <li
                                key={key}
                                className="list-none py-6 border-0 border-solid last:border-b-0"
                            >
                                <figure className="flex flex-col">
                                    <blockquote className="border-solid border-0 border-l-2 border-l-[#4675aa] pl-4">
                                        <p>{post}</p>
                                    </blockquote>
                                    <figcaption className="self-end italic">
                                        - {name} at {postDate}
                                    </figcaption>
                                </figure>
                            </li>
                        );
                    })}
                </ul>
            </section>
        </Layout>
    );
};

export default Feed;
