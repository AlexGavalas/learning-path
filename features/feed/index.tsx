import { useState, useRef, useCallback } from 'react';

import { useUser } from '@lib/use-user';
import { Dialog } from '@components/dialog';
import { Button } from '@components/button';
import { ListItem } from './item';

interface FeedProps {
    posts: Post[];
    onPostDelete: (id: string) => Promise<void>;
    onPostUpdate: (id: string, newPost: string) => Promise<void>;
}

export const Feed = ({ posts, onPostDelete, onPostUpdate }: FeedProps) => {
    const [openDialog, setOpenDialog] = useState(false);
    const idToDelete = useRef<string>();
    const { user } = useUser();

    const closeDialog = useCallback(() => setOpenDialog(false), []);

    const handleDelete = (id: string) => {
        idToDelete.current = id;
        setOpenDialog(true);
    };

    const deleteCurrentId = async () => {
        const id = idToDelete.current;

        if (!id) return;

        await onPostDelete(id);
    };

    if (!posts.length) {
        return <p className="text-center">No posts yet</p>;
    }

    return (
        <>
            <ul className="mt-4">
                {posts.map((post) => {
                    const isMine = post.name === user?.user_metadata.name;

                    return (
                        <ListItem
                            key={post.id}
                            post={post}
                            isMine={true}
                            handleDelete={handleDelete}
                            onPostUpdate={onPostUpdate}
                        />
                    );
                })}
            </ul>
            <Dialog open={openDialog} onClickOutside={closeDialog}>
                <p>Are you sure you want to delete?</p>
                <div className="flex gap-2 justify-end">
                    <Button onClick={closeDialog}>No</Button>
                    <Button
                        onClick={() => {
                            deleteCurrentId();
                            closeDialog();
                        }}
                    >
                        Yes
                    </Button>
                </div>
            </Dialog>
        </>
    );
};
