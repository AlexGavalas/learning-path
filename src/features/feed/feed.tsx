import { useState, useCallback } from 'react';

import { useUser } from '~lib/use-user';
import { Dialog } from '~components/dialog';
import { Button } from '~components/button';
import { ListItem } from './item';

interface FeedProps {
    posts: UserPost[];
    onPostDelete: (id: string) => Promise<void>;
    onPostUpdate: (id: string, newPost: string) => Promise<void>;
}

export const Feed = ({ posts, onPostDelete, onPostUpdate }: FeedProps) => {
    const [openDialog, setOpenDialog] = useState(false);
    const [idToDelete, setIdToDelete] = useState('');
    const [error, setError] = useState<string>();
    const { user } = useUser();

    const closeDialog = useCallback(() => setOpenDialog(false), []);

    const handleDelete = (id: string) => {
        setIdToDelete(id);
        setOpenDialog(true);
    };

    const deleteCurrentId = async () => {
        try {
            await onPostDelete(idToDelete);

            closeDialog();
        } catch (e) {
            const errorMessage =
                e instanceof Error ? e.message : 'An unknown error occured';

            setError(errorMessage);
        }
    };

    if (!posts.length) {
        return <p className="text-center">No posts yet</p>;
    }

    return (
        <>
            <ul className="mt-4">
                {posts.map((post) => {
                    const isMine = post.name === user?.user_metadata.name;

                    return (
                        <ListItem
                            key={post.id}
                            post={post}
                            isMine={isMine}
                            onPostDelete={handleDelete}
                            onPostUpdate={onPostUpdate}
                        />
                    );
                })}
            </ul>
            <Dialog open={openDialog} onClickOutside={closeDialog}>
                <p>Are you sure you want to delete?</p>
                {error && <span>{error}</span>}
                <div className="flex gap-2 justify-end">
                    <Button onClick={closeDialog}>No</Button>
                    <Button onClick={deleteCurrentId}>Yes</Button>
                </div>
            </Dialog>
        </>
    );
};
