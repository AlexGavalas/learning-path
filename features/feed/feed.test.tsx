import { ReactElement } from 'react';
import { render, RenderOptions, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe } from 'jest-axe';
import { UserContextProvider } from '~lib/use-user';

import { Feed } from './feed';

jest.mock('~lib/supabase', () => ({
    supabase: {
        auth: {
            onAuthStateChange: jest.fn().mockReturnValue({
                data: {
                    subscription: {
                        unsubscribe: jest.fn(),
                    },
                },
            }),
            getSession: jest
                .fn()
                .mockResolvedValue({ data: { session: null } }),
        },
    },
}));

const setup = (ui: ReactElement, options?: RenderOptions) => ({
    user: userEvent.setup(),
    ...render(ui, options),
});

const POSTS: Post[] = [
    {
        id: '1',
        created_at: new Date('2022-01-01').toISOString(),
        name: 'name',
        post: 'post',
    },
];

describe('<Feed />', () => {
    const renderFeed = ({ posts }: { posts: Post[] }) => {
        const mockPostDelete = jest.fn();
        const mockPostUpdate = jest.fn();

        const renderResult = setup(
            <Feed
                posts={posts}
                onPostDelete={mockPostDelete}
                onPostUpdate={mockPostUpdate}
            />,
            { wrapper: UserContextProvider },
        );

        return {
            ...renderResult,
            mockPostDelete,
            mockPostUpdate,
        };
    };

    describe('when posts are empty', () => {
        it('renders', () => {
            const { container } = renderFeed({ posts: [] });

            expect(container).toMatchSnapshot();
        });

        it('is accessible', async () => {
            const { container } = renderFeed({ posts: [] });

            const a11yResults = await axe(container);

            expect(a11yResults).toHaveNoViolations();
        });
    });

    describe('when posts exist', () => {
        const { container } = renderFeed({ posts: POSTS });

        expect(container).toMatchSnapshot();
    });
});
