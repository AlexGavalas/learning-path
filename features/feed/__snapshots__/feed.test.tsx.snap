// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div>
  <ul
    class="mt-4"
  >
    <li
      class="list-none py-6 border-0 border-solid last:border-b-0 flex flex-col"
    >
      <p
        class="border-solid border-0 border-l-2 border-l-teal-500 dark:border-l-yellow-500 pl-4 break-words"
      >
        post
      </p>
      <div
        class="self-end flex gap-4 items-center"
      >
        <p
          class="italic text-sm "
        >
          - 
          name
           at 
          01/01/2022 02:00
        </p>
      </div>
    </li>
  </ul>
</div>
`;

exports[`<Feed /> when posts are empty renders 1`] = `
<div>
  <p
    class="text-center"
  >
    No posts yet
  </p>
</div>
`;
